### PR TITLE
Evosuite plugin interface

### LINKED ISSUE #15: Set up the link between EvoSuite and the plugin
# Description
Create basic integration between EvoSuite and the plugin

# Definition of Done
* [ ]  Make it possible to generate tests for a (hardcoded) class from the plugin (i.e. execute the same functionality that a command such as `java -jar evosuite-1.2.0.jar -projectCP target/classes/ -class delft.ArrayUtils` would have)
* [ ] Resolve any issues such as Java version or dependency incompatibilities that would prevent this integration from working

### LINKED ISSUE #25: Use UI for displaying generated test cases after generating tests for a class
# Description
Link the UI from #23 to actual test generation logic as described in #7 

# Definition of Done
- [ ] After generating tests for a class, the tests are shown using the aforementioned UI

### PR DESCRIPTION
**Pull request by lyuben-todorov**
# Description of changes made
* Expanded GenerateTestsAction so that it extracts information about what is the first java class definition in the currently opened file in the IDE. The class FQN is given, which together with the currently hardcoded project build classpath for gradle projects `target/classes` is enough information to run evosuite and make it generate tests for the user's class.
* Added a class that launches evosuite as a separate process with supplied parameters. 
~~* Added a configurable which determines the location of the evosuite jar that is used.~~
* Evosuite jar that is used is shipped with the plugin.

closes #25
closes #15

### CODE REVIEW

**Comment by lyuben-todorov** [comment]
* Added topic for publishing test generation results objects
* Implemented deserialization and publishing of test generation results
* Added project-level event listener for said topic that can be expanded 

Some things are hardcoded and all of this relies on a custom build of evosuite which comes with this PR. Eventually the jar should be moved elsewhere with its own build process.

**Comment by lyuben-todorov** [comment]
* Integrated with #26

File: src/main/kotlin/com/github/mitchellolsthoorn/testgenie/listener/TestGenerationResultListenerImpl.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import com.intellij.openapi.project.Project
+import org.evosuite.utils.CompactReport
+
+class TestGenerationResultListenerImpl(val project: Project) : TestGenerationResultListener {
+    private val log = Logger.getInstance(this.javaClass)
+
+    override fun testGenerationResult(testReport: CompactReport) {
+        log.info("Received test result for " + testReport.UUT)
+
+        // val actionManager: ActionManager = ActionManager.getInstance()
```
Comment (jzelenjak) [inline src/main/kotlin/com/github/mitchellolsthoorn/testgenie/listener/TestGenerationResultListenerImpl.kt:18]: Do you need this comment here?

File: src/main/kotlin/com/github/mitchellolsthoorn/testgenie/toolwindow/TestGenieToolWindow.kt [hunk 2]
```diff
@@ -54,32 +57,72 @@ class TestGenieToolWindow {
...
+        .addVerticalGap(25)
+        .addComponent(panelTitle)
+        .addLabeledComponent(customLabel("Search budget", "Maximum search duration."), searchBudget, 25, false)
+        .addTooltip(default("60 seconds"))
+        .addLabeledComponent(
+            customLabel("Local search budget type", "Interpretation of local search budget value."),
+            localSearchBudgetType,
+            20,
+            false
+        )
```
Comment (jzelenjak) [inline src/main/kotlin/com/github/mitchellolsthoorn/testgenie/toolwindow/TestGenieToolWindow.kt:70]: Why did you refactor the code that is not part of this merge request?

File: src/main/resources/META-INF/plugin.xml [hunk 3]
```diff
@@ -8,17 +8,12 @@
     <!-- add java functionality. See https://blog.jetbrains.com/platform/2019/06/java-functionality-extracted-as-a-plugin/ for more information -->
     <depends>com.intellij.modules.java</depends>
 
-    <description>
-        <![CDATA[
-            <p>A plugin that uses EvoSuite to generate tests. Developed by TU Delft students.
```
Comment (jzelenjak) [inline src/main/resources/META-INF/plugin.xml:13]: Why did you remove these lines? It was just a placeholder of now

File: src/main/kotlin/com/github/mitchellolsthoorn/testgenie/evosuite/EvoSuiteResultWatcher.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.File
+import java.io.FileReader
+
+
+class EvoSuiteResultWatcher(private val project: Project, private val resultPath: String) : Runnable {
+    private val log = Logger.getInstance(EvoSuiteResultWatcher::class.java)
+
+    override fun run() {
+        val sep = File.separatorChar
+        val testResultDirectory = "${FileUtilRt.getTempDirectory()}${sep}testGenieResults${sep}"
```
Comment (jzelenjak) [inline src/main/kotlin/com/github/mitchellolsthoorn/testgenie/evosuite/EvoSuiteResultWatcher.kt:18]: I admit, this has been done very smartly

File: src/main/kotlin/com/github/mitchellolsthoorn/testgenie/toolwindow/TestGenieToolWindowFactory.kt [hunk 5]
```diff
@@ -17,6 +20,14 @@ class TestGenieToolWindowFactory : ToolWindowFactory {
     override fun createToolWindowContent(project: Project, toolWindow: ToolWindow) {
         val testGeniePanelWrapper = TestGenieToolWindow()
         val contentFactory : ContentFactory = ContentFactory.SERVICE.getInstance()
+
+        val testCaseDisplayService = project.service<TestCaseDisplayService>()
+        toolWindow.contentManager.addContent(
+            contentFactory.createContent(testCaseDisplayService.mainPanel,
```
Comment (jzelenjak) [inline src/main/kotlin/com/github/mitchellolsthoorn/testgenie/toolwindow/TestGenieToolWindowFactory.kt:26]: This looks a bit inconsistent with adding content below. Maybe you can change the indentation here to be more consistent with the one below?

**Review by jzelenjak — COMMENTED** [review]
In general looks good to me. Some parts were done very smartly.
I have left some comments on the code.
For the future: please don't modify/refactor the code that is not related to your pull requests, as it can result in merge conflicts and it skews the code contribution.

**Comment by martinmladenov** [comment]
@lyuben-todorov Can you merge development into your branch? There are merge conflicts. Also, please review and fix Codana's warnings.

File: src/main/kotlin/com/github/mitchellolsthoorn/testgenie/evosuite/EvoSuiteResultWatcher.kt [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        val tmpDir = File(testResultDirectory)
+
+        if (!tmpDir.exists()) {
+            tmpDir.mkdirs()
+        }
+
+        log.info("Started result listener thread for $resultPath")
+
+        while (true) {
+            Thread.sleep(5000)
```
[the remark this answers is not in the document]
Comment (lyuben-todorov) [reply src/main/kotlin/com/github/mitchellolsthoorn/testgenie/evosuite/EvoSuiteResultWatcher.kt:29]: No special reason, 5 seconds seemed like a good period to perform a costly poll (filestystem accesses) for the end result of a process that may take quite a bit of time.

File: build.gradle.kts [hunk 7]
```diff
@@ -32,6 +38,9 @@ intellij {
 
     // Plugin Dependencies. Uses `platformPlugins` property from the gradle.properties file.
     plugins.set(properties("platformPlugins").split(',').map(String::trim).filter(String::isNotEmpty))
+
+    plugins.set(listOf("com.intellij.java"))
```
[the remark this answers is not in the document]
Comment (lyuben-todorov) [reply build.gradle.kts:42]: I _think_ this may be a leftover from before I merged with some other changes. I remember not being able to get PSI to work without setting that plugin. Removing it works fine, I'm guessing due to the thing in `gradle.properties`

File: src/main/kotlin/com/github/mitchellolsthoorn/testgenie/toolwindow/TestGenieToolWindow.kt (hunk 2 above)
Comment (lyuben-todorov) [reply src/main/kotlin/com/github/mitchellolsthoorn/testgenie/toolwindow/TestGenieToolWindow.kt:70]: I think the changes here come as a result of me merging #26 prematurely to my branch.

File: src/main/resources/META-INF/plugin.xml (hunk 3 above)
Comment (lyuben-todorov) [reply src/main/resources/META-INF/plugin.xml:13]: I was getting warnings during the build process from the `patchPluginXml` gradle task that I couldn't fix so I removed it for now.

File: src/main/kotlin/com/github/mitchellolsthoorn/testgenie/evosuite/EvoSuiteResultWatcher.kt (hunk 4 above)
Comment (lyuben-todorov) [reply src/main/kotlin/com/github/mitchellolsthoorn/testgenie/evosuite/EvoSuiteResultWatcher.kt:18]: Actually, I find it pretty terrible and I'd want to isolate this chunk of code as soon as possible.

File: src/main/kotlin/com/github/mitchellolsthoorn/testgenie/toolwindow/TestGenieToolWindowFactory.kt (hunk 5 above)
Comment (lyuben-todorov) [reply src/main/kotlin/com/github/mitchellolsthoorn/testgenie/toolwindow/TestGenieToolWindowFactory.kt:26]: Oops, I may have forgotten to format?

File: src/main/resources/META-INF/plugin.xml (hunk 3 above)
Comment (jzelenjak) [reply src/main/resources/META-INF/plugin.xml:13]: Alright

File: src/main/kotlin/com/github/mitchellolsthoorn/testgenie/evosuite/EvoSuiteResultWatcher.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (lyuben-todorov) [reply src/main/kotlin/com/github/mitchellolsthoorn/testgenie/evosuite/EvoSuiteResultWatcher.kt:29]: good idea

**Comment by lyuben-todorov** [comment]
@martinmladenov merged and fixed

**Review by kirilvasilev16 — APPROVED** [review]
Thank you for going over all the suggestions made by the team. I have no explicit suggestions to fix.

**Review by jzelenjak — APPROVED** [review]
